Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -359,6 +359,43 @@ void TerrainSystem::RefreshArea(AZ::EntityId areaId)
     m_terrainHeightDirty = true;
 }
 
+AZ::EntityId TerrainSystem::FindBestLayerSpawnerAt(const AZ::u32 x, const AZ::u32 y) const
+{
+    // Returns the entityId that contains the highest priority layer encompassing the given position.
+    // If more than one layer has the same priority, will return the last one seen.
```

Comment:
I know our current code is far from optimized, but this is going to be a really expensive operation to run N^2 times for priorities that are fixed across a large region.  To avoid this, maybe m_registeredAreas should be changed from an unordered_map to a list that's sorted by priority.  That would allow the height refresh loop to just bail at the first layer spawner that contains it.  

It will make the area refresh slower to look up, but refreshes should happen a _lot_ less frequently than height queries.  (If the area refresh started to become a bottleneck, we could always keep a list _and_ a map)